### PR TITLE
Resource aliases

### PR DESCRIPTION
**Pull request by dantleech**
This PR aims to allow registring aliases for resource _payloads_ and adds support for decorating the REST response.

For example:

``` javascript
            {
                "path": "\/foo",
                "repo_path": "\/foo",
                "children": [],
                "node": {
                    "jcr:primaryType": "nt:unstructured",
                    "jcr:mixinTypes": [
                        "phpcr:managed"
                    ],
                    "phpcr:class": "Symfony\\Cmf\\Bundle\\ResourceRestBundle\\Tests\\Resources\\TestBundle\\Document\\Article",
                    "phpcr:classparents": [],
                    "title": "Article 1",
                    "body": "This is my article"
                },
                "repository_alias": "phpcr_repo",
                "repository_type": "doctrine_phpcr",
                "payload_alias": "nt:unstructured",
                "_links": {
                    "self": {
                        "href": "\/api\/phpcr_repo\/foo"
                    }
                }
            }
                "phpcr:classparents": [],
                "title": "Article 2"
            },
        },
    },
}
```

``` yaml
cmf_resource_rest:
    payload_alias_map:
        article:
            repository: doctrine_phpcr
            type: "Symfony\\Cmf\\Bundle\\ResourceRestBundle\\Tests\\Resources\\TestBundle\\Document\\Article"
```

### CODE REVIEW

File: Serializer/EventSubscriber/ResourceSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * @param PreSerializeEvent $event
+     */
+    public function onPostSerialize(ObjectEvent $event)
+    {
+        $object = $event->getObject();
+
+        if ($object instanceof Resource) {
+            $visitor = $event->getVisitor();
+            $visitor->addData('repository', $this->registry->getName($object->getRepository()));
```
Comment (dantleech) [inline Serializer/EventSubscriber/ResourceSubscriber.php]: This serializer will add the extra data..

**Comment by dantleech** [comment]
Updated, this should be ready now (pending review)

I have added the following:

```
                "repository_alias": "phpcr_repo",
                "repository_type": "doctrine_phpcr",
                "payload_alias": "nt:unstructured",
```

The `payload_alias` is either the node type (in the case of a PHPCR node) or a FQCN in the case of PHPCR-ODM. "Aliases" can be mapped as above to provide sane names. I guess the principle is the same as with XML namespaces.

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     * Returns the config tree builder.
+     *
+     * @return TreeBuilder
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $treeBuilder->root('cmf_resource_rest')
+            ->children()
+                ->arrayNode('payload_alias_map')
```
Comment (wouterj) [inline DependencyInjection/Configuration.php:31]: requires `->fixXmlConfig`

**Comment by wouterj** [comment]
I can't find `Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest\PayloadAliasRegistry` anywhere in this PR (and travis can't either).

File: DependencyInjection/Configuration.php (hunk 2 above)
Comment (dantleech) [inline DependencyInjection/Configuration.php:31]: Does it?

https://github.com/symfony-cmf/ResourceRestBundle/pull/8/files#diff-d89eb38c26b7d8b615be4d9f58b7dc5cR7

**Comment by dantleech** [comment]
Added it. The travis builds are going to fail anyway as there are pending PRs in resource and resource-bundle.

File: Serializer/EventSubscriber/ResourceSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+use JMS\Serializer\EventDispatcher\EventSubscriberInterface;
+use PHPCR\NodeInterface;
+use JMS\Serializer\EventDispatcher\Events;
+use JMS\Serializer\EventDispatcher\PreSerializeEvent;
+use Puli\Repository\Api\ResourceCollection;
+use JMS\Serializer\EventDispatcher\ObjectEvent;
+use Puli\Repository\Api\Resource\Resource;
+use Symfony\Cmf\Component\Resource\RepositoryRegistryInterface;
+use Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest\PayloadAliasRegistry;
```
Comment (dantleech) [inline Serializer/EventSubscriber/ResourceSubscriber.php]: Remove unused statements

File: DependencyInjection/Configuration.php (hunk 2 above)
Comment (wouterj) [inline DependencyInjection/Configuration.php:31]: Yes, that XML is invalid. It is using an undefined `article` element. It has to be:

``` xml
<?xml version="1.0" charset="UTF-8" ?>
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:cmf-resource-rest="http://cmf.symfony.com/schema/dic/cmf_resource_rest"
    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
    <cmf-resource-rest:config>
        <cmf-resource-rest:payload-alias-map alias="article"
            type="Namespace\Article"
            repository="doctrine_phpcr_odm"
        />
    </cmf-resource-rest:config>
</container>
```

File: DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                             ->scalarNode('type')->end()
+                         ->end()
+                    ->end()
+                ->end()
+            ->end();
+
+        return $treeBuilder;
+    }
+}
+
```
Comment (wouterj) [inline DependencyInjection/Configuration.php]: empty line

File: ResourceRest/PayloadAliasRegistry.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+/*
+ * This file is part of the Symfony CMF package.
+ *
+ * (c) 2011-2014 Symfony CMF
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest;
```
Comment (wouterj) [inline ResourceRest/PayloadAliasRegistry.php]: I'm -1 for such duplicated namespaces, we should move this to the root of the bundle instead in that case imo

File: ResourceRest/PayloadAliasRegistry.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * @param RepositoryRegistryInterface $repositoryRegistry
+     * @param array $aliases
+     */
+    public function __construct(
+        RepositoryRegistryInterface $repositoryRegistry,
+        array $aliases = array()
+    )
+    {
```
Comment (wouterj) [inline ResourceRest/PayloadAliasRegistry.php]: this has to be `) {`

File: ResourceRest/PayloadAliasRegistry.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+        $type = $resource->getPayloadType();
+
+        if (null === $type) {
+            return null;
+        }
+
+        if (!isset($this->aliasesByRepository[$repositoryType])) {
+            throw new \RuntimeException(sprintf(
+                'No repositories registered with alias "%s". I know about "%s"',
```
Comment (wouterj) [inline ResourceRest/PayloadAliasRegistry.php]: "I know about" -> "Known aliases:"

File: ResourceRest/PayloadAliasRegistry.php (hunk 5 above)
Comment (dantleech) [inline ResourceRest/PayloadAliasRegistry.php]: Well, I do this with the aim of creating a component .. but in this case we could move it to the bundle root.

File: DependencyInjection/Configuration.php (hunk 2 above)
Comment (dantleech) [inline DependencyInjection/Configuration.php:31]: Hmm. Invalid why? If we change it then this would just be `alias` .. how would that look in YAML?

Comment (wouterj) [inline DependencyInjection/Configuration.php:31]: Yaml would be the same. The `->useAttributeAsKey` will transform the `alias` attribute in XML to the key of the array. In Symfony XML, prototyped arrays look like this:

``` xml
<alias id="article" type="..." />
<alias id="other_alias" type="..." />
```

``` yaml
alias_map:
    article: { type: '...' }
    other_alias: { type: '...' }
```

``` php
$root
    ->fixXmlConfig('alias', 'alias_map')
    ->children()
        ->arrayNode('alias_map')
            ->useAttributeAsKey('id')
            ->prototype('array')
                ->children()
                    ->scalarNode('type')->end()
                ->end()
            ->end()
        ->end() // alias_map
    ->end();
```

File: ResourceRest/PayloadAliasRegistry.php (hunk 5 above)
Comment (wouterj) [inline ResourceRest/PayloadAliasRegistry.php]: Would it be usefull to extract a component out of this bundle?

Comment (dantleech) [inline ResourceRest/PayloadAliasRegistry.php]: I don't think so.

File: Decorator/DecoratorInterface.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /**
+     * Decorate the given serialization context.
+     *
+     * For example:
+     *
+     *     $context->addData('foobar', 'Some value');
+     *
+     * @param Context Serialization context
+     * @param resource The resource being serialized
```
Comment (dantleech) [inline Decorator/DecoratorInterface.php]: Resource

File: DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                    ->useAttributeAsKey('name')
+                    ->prototype('array')
+                         ->children()
+                             ->scalarNode('repository')->end()
+                             ->scalarNode('type')->end()
+                         ->end()
+                    ->end()
+                ->end()
+                ->arrayNode('decorator_map')
+                    ->useAttributeAsKey('repository')
```
Comment (dantleech) [inline DependencyInjection/Configuration.php]: fix XML config

File: Tests/Unit/DependencyInjection/fixtures/config.xml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:cmf_resource_rest="http://cmf.symfony.com/schema/dic/cmf_resource_rest"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
+        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+    <cmf_resource_rest:config>
+        <cmf_resource_rest:payload_alias name="article" type="Namespace\Article" repository="doctrine_phpcr_odm" />
+    </cmf_resource_rest:config>
```
Comment (dantleech) [inline Tests/Unit/DependencyInjection/fixtures/config.xml:9]: Add decorator map test

**Comment by dantleech** [comment]
Updated to include support for decorating the response. In doing so I removed the `payload` element from the response and made it into a decorator.

Next step I think will be making a SonataAdmin decorator.

File: Decorator/PayloadDecorator.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ */
+class PayloadDecorator implements DecoratorInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function decorate(Context $context, Resource $resource)
+    {
+        $visitor = $context->getVisitor();
+        $visitor->addData('payload', $context->accept($resource->getPayload()));
```
Comment (dantleech) [inline Decorator/PayloadDecorator.php]: @WouterJ @dbu is this even decoration? or should we use a different term?

Comment (dantleech) [inline Decorator/PayloadDecorator.php]: I guess the Enhancer might be better in fitting with the RoutingBundle.

**Comment by dantleech** [comment]
Test failing due to deprecation errors, see: https://github.com/symfony-cmf/Testing/pull/93

**Pull request merged**